### PR TITLE
init project and init tracing API

### PR DESCRIPTION
**Pull request by ascrutae**

### CODE REVIEW

**Comment by ascrutae** [comment]
@wu-sheng I had finished the trace api, Could you help me to check this PR if you are free.

File: README.md [hunk 1]
```diff
@@ -1,2 +1,2 @@
 # skywalking-nodejs
-The Node server instrument SDK for Apache SkyWalking
+The Node serverB instrument SDK for Apache SkyWalking
```
Comment (wu-sheng) [inline README.md]: Please make sure the README is official and right.

File: .gitignore [hunk 2]
```diff
@@ -57,3 +57,9 @@ typings/
 # dotenv environment variables file
 .env
 
+npm-debug.log
+.DS_Store
+.idea/
+
+package-lock.json
+example/
```
Comment (wu-sheng) [inline .gitignore]: Why does git ignore include the `example` folder?

File: lib/plugins/supported/http.js [hunk 3]
```diff
@@ -0,0 +1,69 @@
...
+
+    instrumentation.enhanceMethod(httpModule, 'request', wrapRequest);
+
+    return httpModule;
+
+    function wrapEmit(original) {
+        return function wrappedEmit(event, req, res) {
+            if (event === 'request') {
+                var contextCarrier = undefined;
+                if (req.headers.hasOwnProperty("sw3")) {
```
Comment (wu-sheng) [inline lib/plugins/supported/http.js]: The ContextCarrier API is not coded in right way.

File: lib/instrumentation/index.js [hunk 4]
```diff
@@ -0,0 +1,102 @@
...
+ * @returns {*}
+ */
+Instrumentation.prototype.enhanceCallback = function (traceContext, contextManager, originCallBack) {
+    if (typeof originCallBack !== 'function') return originCallBack
+
+    var runningTraceContext = traceContext;
+
+    return function () {
+        var previousTraceContext = contextManager.activeTraceContext();
+        contextManager.active(runningTraceContext);
```
Comment (wu-sheng) [inline lib/instrumentation/index.js]: Is there any possibility the runningTraceContext or previousTraceContext `unidentified`?

File: lib/instrumentation/index.js [hunk 5]
```diff
@@ -0,0 +1,102 @@
...
+
+        try {
+            var result = originCallBack.apply(this, arguments);
+        } catch (err) {
+            runningTraceContext.errorOccurred();
+            runningTraceContext.log(err);
```
Comment (wu-sheng) [inline lib/instrumentation/index.js]: Why could the context log and tag error?

File: lib/trace/context-manager.js [hunk 6]
```diff
@@ -0,0 +1,136 @@
...
+/**
+ * @author zhang xin
+ */
+function ContextManager() {
+    this._activeTraceContext = undefined;
+    this._dictionaryManager = new DictionaryManager();
+}
+
+ContextManager.prototype.createSpan = function (spanOptions) {
+    if (!AgentConfig.getApplicationId()) {
```
Comment (wu-sheng) [inline lib/trace/context-manager.js]: When  application instance id is NULL, span shouldn't be created.

File: lib/trace/context-manager.js [hunk 7]
```diff
@@ -0,0 +1,136 @@
...
+
+    if (contextCarrier) {
+        span.traceContext().extract.apply(span.traceContext(), [contextCarrier, this._agentConfig]);
+    }
+
+    return span;
+};
+
+ContextManager.prototype.createExitSpan = function (operationName, peerId, contextCarrier) {
+    if (!AgentConfig.getApplicationId()) {
```
Comment (wu-sheng) [inline lib/trace/context-manager.js:104]: When  application instance id is NULL, span shouldn't be created.

Comment (wu-sheng) [reply lib/trace/context-manager.js:104]: And `#createSpan` has already included this check, where do you prefer to do this?

**Review by wu-sheng — CHANGES_REQUESTED** [review]
I have some concerns in comments. Besides that, I also want to know, what happens if TraceContext works with some NoopContext? Because from your codes, the context created with span, and application id and instance id could be set in any time.

File: lib/plugins/supported/http.js [hunk 8]
```diff
@@ -0,0 +1,73 @@
...
+
+    instrumentation.enhanceMethod(httpModule, 'request', wrapRequest);
+
+    return httpModule;
+
+    function wrapEmit(original) {
+        return function wrappedEmit(event, req, res) {
+            if (event === 'request') {
+                var contextCarrier = new ContextCarrier();
+                contextCarrier.fetchTracingContext(function (key) {
```
Comment (wu-sheng) [inline lib/plugins/supported/http.js]: Why named `#contextCarrier.fetchTracingContext`? In our module, this should be named as `#extract`

**Review by wu-sheng — APPROVED** [review]

**Pull request merged**